deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/Layer.java

```diff
@@ -278,8 +301,29 @@ public T constraints(LayerConstraint... constraints) {
          *
          * @param constraints Constraints to apply to all layers
          */
-        public T constraints(List<LayerConstraint> constraints) {
+        public T weightConstraints(LayerConstraint... constraints) {
+            return constraints(Arrays.asList(constraints), false,
+                    true, null);
+        }
+
+        /**
+         * Set constraints to be applied to this layer. Default: no constraints.<br>
+         * Constraints can be used to enforce certain conditions (non-negativity of parameters, max-norm regularization,
+         * etc). These constraints are applied at each iteration, after the parameters have been updated.
+         *
+         * @param constraints Constraints to apply to this layer
+         * @param hasBiasConstraints if contraints will be applied to bias parameters
+         * @param hasWeightConstraints if constraints will be applied to weight parameters
+         * @param constraintParamNames Set of parameter names that constraints will be applied to. Overrides the two
+         *                             boolean flags before.
+         * @return Constraints
+         */
+        public T constraints(List<LayerConstraint> constraints, boolean hasBiasConstraints,
+                             boolean hasWeightConstraints, Set<String> constraintParamNames) {
```

Comment:
This approach won't work: we can't use ```weightConstraints(x).biasConstraints(y)``` as the latter will override the former (i.e., x will be removed entirely, and only y will remain).
And again, the weight vs. bias vs. recurrent weight selection should IMO be in the constraints themselves, not as fields on the layer.